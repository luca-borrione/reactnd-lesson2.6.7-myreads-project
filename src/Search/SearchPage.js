import React from 'react';
import { Link } from 'react-router-dom';
import SearchBar from './SearchBar';
import PanelError from '../PanelError';
import BooksGrid from '../shared/BooksGrid';
import PropTypes from 'prop-types';

/**
 * @module
 * @name SearchPage
 * @extends React.Component
 * @description
 * Contains the component [SearchBar]{@link module:SearchBar} to perform a search
 * for new books and a Link to navigate back to the [MyReads]{@link module:MyReads} component.
 * If an error has been triggered during the search, it will show the component
 * [PanelError]{@link module:PanelError}, otherwise it will order the books retrieved
 * by the search using the [BooksGrid]{@link module:BooksGrid} component.
 */
class SearchPage extends React.Component {

	/**
	 * @member
	 * @name propTypes
	 * @description Intended types of the props passed to the component
	 * @property {TBook[]} getBookShelf - see [getBookShelf]{@link module:App~getBookShelf}
	 * @property {function} updateBookShelf - see [updateBookShelf]{@link module:App~updateBookShelf}
	 * @static
	 */
	static propTypes = {
		getBookShelf: PropTypes.func.isRequired,
		updateBookShelf: PropTypes.func.isRequired
	};


	constructor(props) {
		super(props);
		this.showResult = this.showResult.bind(this);
	}


	/**
	 * @member
	 * @name state
	 * @property {TBook[]} books - List of the books retrieved by the search
	 * @private
	 */
	state = {
		books: []
	};


	/**
	 * @method
	 * @name showResult
	 * @description Shows the books retrieved by a search
	 * @param {TBook[]} books
	 */
	async showResult(books) {
		await this.setState({ books });
	}


	/**
	 * @method
	 * @name render
	 * @description
	 * Renders
	 * - A Link to navigate back to the [MyReads]{@link module:MyReads} component
	 * - The [SearchBar]{@link module:SearchBar} to perform a search for new books
	 * - The [PanelError]{@link module:PanelError} if an error has been triggered
	 * during the search, or the [BooksGrid]{@link module:BooksGrid} component otherwise.
	 * @returns {ReactElement}
	 * @private
	 */
	render() {
		const { BOOKS_STATUS } = SearchBar;
		const { books } = this.state;
		const { getBookShelf, updateBookShelf } = this.props;

		return (
			<div className="search-books">
				<div className="search-books-bar">
					<Link to='/' id="goto-home" className="close-search">Close</Link>
					<SearchBar
						getBookShelf={getBookShelf}
						showResult={this.showResult} />
				</div>

				{books.status === BOOKS_STATUS.ERROR
					? (
						<PanelError />
					)
					: (
						<div className="search-books-results">
							<BooksGrid
								books={books}
								updateBookShelf={updateBookShelf}  />
						</div>
					)}
			</div>
		);
	}
}

export default SearchPage;
