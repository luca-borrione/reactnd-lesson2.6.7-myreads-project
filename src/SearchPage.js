import React from 'react';
import { Link } from 'react-router-dom';
import SearchBar from './SearchBar';
import PanelError from './PanelError';
import BooksGrid from './BooksGrid';
import PropTypes from 'prop-types';

/**
 * @class SearchPage
 * @extends React.Component
 * @classdesc
 * Allows the user to perform a search for new books
 * @hideconstructor
 */
class SearchPage extends React.Component {

	/**
	 * @property {Object} propTypes - Intended types passed to the component
	 * @property {function} propTypes.getBookShelf
	 * 	Method of the [App]{@link App#getBookShelf} component for retrieving the shelf key
	 * 	associated with a book by book id.
	 * @property {function} propTypes.updateBookShelf
	 * 	Method of the App]{@link App#updateBookShelf} component for updating the list
	 * 	of books in the shelves
	 * @static
	 */
	static propTypes = {
		getBookShelf: PropTypes.func.isRequired,
		updateBookShelf: PropTypes.func.isRequired
	};


	constructor(props) {
		super(props);
		this.showResult = this.showResult.bind(this);
	}

	/**
	 * @property {Object} state
	 * @property {TBook[]} state.books - List of books resulting from the search
	 * @property {string} state.searchedValue - SearchPage keywords typed by the user in the textfield
	 * @private
	 */
	state = {
		books: []
	};


	async showResult(books) {
		await this.setState({ books });
	}


	/**
	 * @description
 	 * - Shows a textfield in which the user can type keywords to search books.<br>
 	 * - Lists the books resulting from the search undernit the textfield by rendering
	 * the [BooksGrid]{@link BooksGrid} component
	 * @returns {ReactElement}
	 */
	render() {
		const { BOOKS_STATUS } = SearchBar;
		const { books } = this.state;
		const { getBookShelf, updateBookShelf } = this.props;

		return (
			<div className="search-books">
				<div className="search-books-bar">
					<Link to='/' id="goto-home" className="close-search">Close</Link>
					<SearchBar
						getBookShelf={getBookShelf}
						showResult={this.showResult} />
				</div>

				{books.status === BOOKS_STATUS.ERROR
					? (
						<PanelError />
					)
					: (
						<div className="search-books-results">
							<BooksGrid
								books={books}
								updateBookShelf={updateBookShelf}  />
						</div>
					)}
			</div>
		);
	}
}

export default SearchPage;
