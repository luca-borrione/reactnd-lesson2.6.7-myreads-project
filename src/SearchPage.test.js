import ReactDOM from 'react-dom';
import { MemoryRouter } from 'react-router';
import renderer from 'react-test-renderer';
import SearchPage from './SearchPage';
import { shallow } from 'enzyme';

describe('SearchPage', () => {

	// beforeEach(() => {
	// 	jest.resetModules();
	// });

	// afterEach(() => {
	// 	jest.clearAllMocks();
	// 	jest.restoreAllMocks();
	// });

	const getBookShelf = () => {};
	const updateBookShelf = () => {};

	it('renders without crashing', () => {
		const div = document.createElement('div');
		ReactDOM.render(
			<MemoryRouter>
				<SearchPage
					getBookShelf={getBookShelf}
					updateBookShelf={updateBookShelf} />
			</MemoryRouter>, div);
		ReactDOM.unmountComponentAtNode(div);
	});


	it('renders correctly', () => {
		const tree = renderer.create(
			<MemoryRouter>
				<SearchPage
					getBookShelf={getBookShelf}
					updateBookShelf={updateBookShelf} />
			</MemoryRouter>
		)
		.toJSON();

		expect(tree).toMatchSnapshot();
	});


	it("includes link to '/'", () => {
		const wrapper = shallow(
			<SearchPage
				getBookShelf={getBookShelf}
				updateBookShelf={updateBookShelf} />
		);

		const link = wrapper.find('Link');
		expect(link.prop('to')).toBe('/');
	});
});
