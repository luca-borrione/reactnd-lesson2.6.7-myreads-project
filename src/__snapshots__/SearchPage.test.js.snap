// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage renders correctly 1`] = `
<div
  className="search-books"
>
  <div
    className="search-books-bar"
  >
    <a
      className="close-search"
      href="/"
      id="goto-home"
      onClick={[Function]}
    >
      Close
    </a>
    <div
      className="search-books-input-wrapper"
    >
      <input
        onChange={[Function]}
        placeholder="Search by title or author"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="search-books-results"
  >
    <ol
      className="books-grid"
    />
  </div>
</div>
`;
